### PR TITLE
Enable user to zoom beyond image size

### PR DESCRIPTION
**Pull request by jaiwanth-v**
Fixes vector-im/element-web/issues/17114

ezgif com-gif-maker (16)

Also added some cleanup functions to the image view.

Signed-off-by: Jaiwanth <jaiwanth2011@gmail.com>

---
Here's what your changelog entry will look like:

## ✨ Features
 * Enable user to zoom beyond image size ([\#5949](https://github.com/matrix-org/matrix-react-sdk/pull/5949)). Contributed by @jaiwanth-v.

### CODE REVIEW

File: src/components/views/elements/ImageView.tsx [hunk 1]
```diff
@@ -114,6 +114,8 @@ export default class ImageView extends React.Component<IProps, IState> {
 
     componentWillUnmount() {
         this.focusLock.current.removeEventListener('wheel', this.onWheel);
+        window.removeEventListener("resize", this.calculateZoom);
+        this.image.current.removeEventListener("load", this.calculateZoom);
```
Comment (SimonBrandner) [inline src/components/views/elements/ImageView.tsx:118]: Nice catch - thanks!

File: src/components/views/elements/ImageView.tsx [hunk 2]
```diff
@@ -159,8 +161,8 @@ export default class ImageView extends React.Component<IProps, IState> {
             });
             return;
         }
-        if (newZoom >= this.state.maxZoom) {
-            this.setState({zoom: this.state.maxZoom});
+        if (newZoom >= 2 * this.state.maxZoom) {
+            this.setState({zoom: 2 * this.state.maxZoom});
```
Comment (SimonBrandner) [inline src/components/views/elements/ImageView.tsx:165]: Could you perhaps export this into a const and explain why are we multiplying by 2?

**Review by SimonBrandner — CHANGES_REQUESTED** [review]
Thanks for working on this!

Peek 2021-04-30 07-41

It seems like this doesn't play nice with images that are smaller than the viewport - the cursor icon always seems to be the same which isn't ideal. Could you please fix that?

(maybe wait for a response from design, so that you don't have to change things twice)

File: src/components/views/elements/ImageView.tsx (hunk 2 above)
Comment (jaiwanth-v) [reply src/components/views/elements/ImageView.tsx:165]: Sure, I'll add a comment.

**Comment by jaiwanth-v** [comment]
Okay, I'll modify that once we get a response from the design team.

**Review by jryans — COMMENTED** [review]
Hmm, this is seems to conflict with Design-provided guidance which triggered https://github.com/vector-im/element-web/issues/17049, since clicking on the image now zooms beyond 100%.

In any case, let's see what Design says on this.

**Comment by SimonBrandner** [comment]
Currently, I think it would make most sense if clicking the zoomed out image zoomed you in to 100% but you would be allowed to zoom in further using the buttons or the scroll wheel. That way most of the old behavior is preserved. That being said, this is just my opinion

**Comment by jryans** [comment]
> Currently, I think it would make most sense if clicking the zoomed out image zoomed you in to 100% but you would be allowed to zoom in further using the buttons or the scroll wheel. That way most of the old behavior is preserved.

That sounds reasonable to me, but I also have no idea what Design thinks of this change, so they may have an entirely different idea in mind.

File: src/components/views/elements/ImageView.tsx (hunk 1 above)
Comment (jryans) [reply src/components/views/elements/ImageView.tsx:118]: We may want to move this listener cleanup to a separate PR... We should definitely land this listener fix sooner, and it's unclear to me when / if the behaviour changes in the rest of this PR would land.

Comment (jaiwanth-v) [reply src/components/views/elements/ImageView.tsx:118]: Okay, I'll make a PR for that soon.

**Comment by niquewoodhouse** [comment]
How can I test this deployment for design review please?

**Comment by SimonBrandner** [comment]
@jaiwanth-v, could you please merge develop?

**Comment by jaiwanth-v** [comment]
I thought this change was no longer required due to some other PR. Am I wrong?

**Comment by SimonBrandner** [comment]
> I thought this change was no longer required due to some other PR. Am I wrong?

From my testing it doesn't seem I can zoom to more than 100% - I don't think there was a PR for this already?

**Comment by niquewoodhouse** [comment]
I'm happy to review on behalf of design, if there's a deployment.  @jaiwanth-v would you mind merging so I can review please?

**Comment by jaiwanth-v** [comment]
Sure. I'll do it soon.

**Comment by andybalaam** [comment]
@niquewoodhouse Travis just pressed the button, so there should be a testable build within 30 minutes or so. Please ping me if it doesn't appear and we can try to fix it.

**Review by niquewoodhouse — APPROVED** [review]

**Review by turt2live — APPROVED** [review]
seems reasonable - thanks!

**Pull request merged**